### ISSUE TITLE
creating an api-group overlay breaks everything

### ISSUE THREAD

**Issue by ralphvanetten** [DapperDox/dapperdox#40]
If I create an API group overlay in the example and run it, I get request timeouts and nothing in the browser when opening http://localhost:3123/swagger-petstore/reference/everything-about-your-pets/get-pet-by-id

If I remove the API group overlay file everything works as expected.

To reproduce:

Create a file `examples/overlay/assets/sections/swagger-petstore/templates/reference/everything-about-your-pets.md` containing:

```
Overlay: true

[[description]]
Hello
```

Update `run_example.sh` to include `-assets-dir=./examples/overlay/assets` and run it:

```
$ ./run_example.sh 
2017/02/09 15:38:09 DapperDox server version 1.1.0 starting
2017/02/09 15:38:09 Configuration:
2017/02/09 15:38:09 	          BindAddr: 0.0.0.0:3123
2017/02/09 15:38:09 	         AssetsDir: ./examples/overlay/assets
2017/02/09 15:38:09 	  DefaultAssetsDir: assets
2017/02/09 15:38:09 	           SpecDir: examples/specifications/petstore/
2017/02/09 15:38:09 	      SpecFilename: [/swagger.json]
2017/02/09 15:38:09 	             Theme: default
2017/02/09 15:38:09 	          ThemeDir: 
2017/02/09 15:38:09 	          LogLevel: info
2017/02/09 15:38:09 	           SiteURL: http://localhost:3123
2017/02/09 15:38:09 	    SpecRewriteURL: [petstore.swagger.io=PETSTORE.swagger.io]
2017/02/09 15:38:09 	DocumentRewriteURL: []
2017/02/09 15:38:09 	     ForceSpecList: %!s(bool=false)
2017/02/09 15:38:09 	        ShowAssets: %!s(bool=true)
2017/02/09 15:38:09 	         ProxyPath: []
2017/02/09 15:38:09 	    TLSCertificate: 
2017/02/09 15:38:09 	            TLSKey: 
2017/02/09 15:38:09 listening on 0.0.0.0:3123
2017/02/09 15:38:09 Registering specifications
2017/02/09 15:38:09 [XVlBzgbaiCMRAjWwhTHc] [info] GET /swagger.json (200, 233.893µs)
2017/02/09 15:38:09 Registering reference documentation
2017/02/09 15:38:09 Registering guides
2017/02/09 15:38:09 listening on 0.0.0.0:3123 for unsecured connections
2017/02/09 15:38:17 [tcuAxhxKQFDaFpLSjFbc] [warn] request timed out
2017/02/09 15:38:46 [XoEFfRsWxPLDnJObCsNV] [warn] request timed out
2017/02/09 15:38:54 [lgTeMaPEZQleQYhYzRyW] [warn] request timed out
```

**Comment by zxchris** [comment]
Hi Ralph.

Running DapperDox with `-assets-dir=./examples/overlay/assets` should pick up a whole load of other overlays in addition to your example. I duplicated your example above in `examples/overlay2/assets` to keep it out of the way, and I get the text `Hello` displayed underneath the heading `Everything about your pets`.

I'm not sure what the difference can be. I'll grab the binary release of v1.1.0 and try that.

In the meantime, what version of DapperDox are you using, and have you built from source or the binary release?

**Comment by ralphvanetten** [comment]
I tried the binary release and that one works fine. 
If I build it myself I get those timeouts.
I guess it has something to do with my go setup.

**Comment by zxchris** [comment]
What version of go are you using?
The binary releases of v1.1.0 were built using `go version go1.7.3 darwin/amd64`

I will be upgrading to the go v1.8 over the next few days, now that it's out (just to warn you that we've not built with v1.8 yet!)

**Comment by ralphvanetten** [comment]
I am using `go version go1.7.4 linux/amd64`

**Comment by zxchris** [comment]
Can you confirm whether your build is working now, so I can close this issue?

Thanks!

**Comment by ralphvanetten** [comment]
I've grabbed the master today and build it but if gives the same issue if I build it myself.
The binary distribution works fine though so it must be something with my local go version which is the standard Debian version.

**Comment by zxchris** [comment]
Is it possible to try on a different machine? We've had a number of people try and reproduce this but have not been able to yet (Centos, Mac OS X and Windows). I'm running Debian Stretch on one machine at home, so will try a build there myself there.

**Comment by ralphvanetten** [comment]
It must be something weird with my machine then. Lets close it.

**Comment by ralphvanetten** [comment]
Sorry for reopening so soon but I would like to change something in the code and therefore I would like to be able to compile it.

I did some more tests and they all gave the same request timed out issue. I tried the official go distributions from golang.org version 1.7.3 and 1.8.1 and I tested it on a fresh debian installation.
I did not test it on a different distribution yet.

If I run my self build dapperdox using the overlay mentioned earlier, I can open the dapperdox documentation in my browser (I can see the front page) but as soon as I follow a link, I get request timeout. After this timeout not even the front page works anymore:

```
2017/05/11 14:23:39 [XoEFfRsWxPLDnJObCsNV] [trace] request completed: 2017-05-11 14:23:39.475867013 +0200 CEST
2017/05/11 14:23:39 [XoEFfRsWxPLDnJObCsNV] [info] GET /swagger-petstore/reference (200, 2.653306ms)
2017/05/11 14:23:42 [lgTeMaPEZQleQYhYzRyW] [trace] request started: 2017-05-11 14:23:42.235864883 +0200 CEST
2017/05/11 14:23:42 -- template: method  Version latest
2017/05/11 14:23:42 Overlay: Looking for overlay banner
2017/05/11 14:23:42 Overlay: Does 'swagger-petstore/templates/reference/everything-about-your-pets/get-pet-by-id/banner/overlay' exist?
2017/05/11 14:23:42 Applying overlay 'swagger-petstore/templates/reference/everything-about-your-pets/get-pet-by-id/banner/overlay'
2017/05/11 14:23:43 [lgTeMaPEZQleQYhYzRyW] [trace] request timed out
2017/05/11 14:23:43 [lgTeMaPEZQleQYhYzRyW] [trace] headers not written, calling failure handler
2017/05/11 14:23:43 [lgTeMaPEZQleQYhYzRyW] [warn] request timed out
```

I will see if I can test it on another distribution but I want to ask if there can be some mistake in the way I compile dapperdox?

I used to get `no install location for directory outside GOPATH` but after rearranging the directory layout and setting GOPATH I got rid of them. What is the recommended directory layout for placing the dapperdox sources?

Thanks.

**Comment by cpage-nexway** [comment]
Hello, I have the same issue when I build dapperdox.
go version: go version go1.8 linux/amd64.

**Comment by bircow** [comment]
I also can confirm this issue exists. I tried to compile Dapperdox 1.1.1 with Go from 1.7.3 to 1.8.3 and with all of them I had "request timed out" if assets contains overlays. All tests were done via Docker Go images which are based on Debian Jessie. If others claim that 1.7.3 don't cause this issue, it can be somehow related to Debian Jessie...

**Comment by ralphvanetten** [comment]
It is not a Debian issue. I've tried it on fedora (25 in a vm, go version 1.7.5 linux/amd64) and it still gives the same issue.

What can be wrong? I really would like to make changes to dapperdox.

Thanks.

**Comment by zxchris** [comment]
We've been able to reproduce this. Finally!
So, we're investigating - the only thing that appears to be different is the version of go that was used to build DapperDox (1.8.x).

**Comment by danbars** [comment]
Hello, Have you been able to resolve this?
I'm having the same problem on Sierra, go ver 1.8.3
I've built dapperdox myself, and also had to manually set GOPATH (pointed to $HOME) and GOBIN (pointed to $HOME/go/bin) to make go work and avoid `no install location for directory outside GOPATH` error on `go install`

**Comment by sergeymisura** [comment]
I ran into the same issue using Go 1.7.3

I did some digging and it seems like I have found the culprit:
* When DapperDox processes an HTML template it calls Render.HTML method from unrolled/render library
* If the render then needs to look up an overlay, it calls ```func overlay()``` from dapperdox/render/render.go
* This function, in turn, calls Render.HTML again to process the overlay template

So essentially we have Render.HTML called again, while the first call to Render.HTML has not been finished yet.

Now, the very first line of the code in this function attempts to lock a mutex
```go
func (r *Render) HTML(w io.Writer, status int, name string, binding interface{}, htmlOpt ...HTMLOptions) error {
    r.templatesLk.Lock()
    defer r.templatesLk.Unlock()
```

which works fine the first time around, but causes a deadlock (and, consequentially, a timeout) when called for the second time.

-------------------------

Bottom line, it looks like the difference is not the version of Go, but the version of the rendering library.

If someone could help me to figure out how to switch to the older version of the library, I'd appreciate it very much - I can't think of an easy fix for this issue at the moment.

**Comment by taxcloud** [comment]
We have rolled dapperdox into production using v.1.2.2 Windows AMD64 and getting this timeout issue consistently (not immediately but after about 30 to 90 minutes) since introducing use of overlays.
Restarting dapperdox resolves the issue for another 30 to 90 minutes. Any suggestions?
